src/elfcore.cc

```diff
@@ -711,30 +721,151 @@ static Ehdr *SanitizeVDSO(Ehdr *ehdr, size_t start, size_t end) {
+  NO_INTR(io_status.fd = sys_open(proc_path, O_RDONLY, 0));
+  if (io_status.fd >= 0) {
+    if (GetChar(&io_status) > 0) {
+      ind = (char *)io_status.buf;
+      key_start = ind;
+      while (ind < (char *)io_status.end) {
+        if (*ind == ':') {
+          *ind = '\0';
+          value_start = ind + 2; /* To skip the :<tab> */
+        }
+        if (*ind == '\n') {
+          *ind = '\0';
+          UpdatePsInfo(prinfo, key_start, value_start);
+          key_start = ++ind;
+        }
+        ind++;
+      }
+    }
+  }
+}
+
+/*
+ * This function helps to find, if we can access the given
+ * address from a process memory particularly vdso section.
+ */
+static int
+PeekVDSO(void *addr, void *dest, size_t size, pid_t pid)
+{
+  int mem_fd, ret = -1;
+  char proc_path[80];
+  size_t nread = 0;
+
+  sprintf(proc_path, "/proc/%d/mem", pid);
+  mem_fd = sys_open(proc_path, O_RDONLY, 0);
+  if (mem_fd == -1) {
+#ifdef DEBUG
+    fprintf(stderr, "proc mem open failed\n");
+#endif
+    return -1;
+  }
+
+  if (sys_lseek(mem_fd, (size_t)addr, SEEK_SET) < 0) {
+#ifdef DEBUG
+    fprintf(stderr, "VDSO seek failed\n");
+#endif
+    goto done;
+  }
+
+  NO_INTR(nread = sys_read(mem_fd, dest, size));
+  if (nread != size) {
+    goto done;
+  } 
+  ret = 0;
+done:
+  sys_close(mem_fd);
+  
+  return ret;
+}
+
 /* This function is invoked from a separate process. It has access to a
  * copy-on-write copy of the parents address space, and all crucial
  * information about the parent has been computed by the caller.
  */
 static int CreateElfCore(void *handle, ssize_t (*writer)(void *, const void *, size_t), int (*is_done)(void *),
                          prpsinfo *prpsinfo, core_user *user, prstatus *prstatus, int num_threads, pid_t *pids,
                          regs *regs, fpregs *fpregs, fpxregs *fpxregs, size_t pagesize, size_t prioritize_max_length,
-                         pid_t main_pid, const struct CoredumperNote *extra_notes, int extra_notes_count) {
+                         pid_t main_pid, const struct CoredumperNote *extra_notes, int extra_notes_count, bool ext_process) {
   /* Count the number of mappings in "/proc/self/maps". We are guaranteed
    * that this number is not going to change while this function executes.
    */
   int rc = -1, num_mappings = 0;
+  int mem_fd = -1;
   struct io io;
+/*
+ * The loopback is used to validate accessiblity of an address, for self process
+ */
   int loopback[2] = {-1, -1};
+  char proc_path[80];
   size_t num_auxv;
   union {
-    Ehdr *ehdr;
+    Ehdr ehdr;
     size_t address;
   } vdso;
 
-  if (sys_pipe(loopback) < 0) goto done;
-
+  if (!ext_process) {
+    if (sys_pipe(loopback) < 0) goto done;
+  }
+  vdso.address = 0;
   io.data = io.end = 0;
-  NO_INTR(io.fd = sys_open("/proc/self/maps", O_RDONLY, 0));
+
+  /*
+   * "/proc/[pid]/maps" should work for both self and external process,
+   * since we don't want any overhead of sprintf in self, we are using
```

Comment:
it's not overhead but avoiding use of library calls, isn't it?

Reply:
I think you should add a util function that simply appends /proc/ + pid + maps; using local_atoi,
then use it in all these places.

Reply:
Done, See my above response.